### PR TITLE
[Blog Draft] LiveView Design Patterns: LiveComponent and the Single Responsibility Principle

### PR DESCRIPTION
**Pull request by SophieDeBenedetto**
After having a chance to play around with LiveComponent, I wrote up this post to share what I learned.

### CODE REVIEW

File: _posts/2019-12-12-live-view-live-component.md [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+date: 2019-10-20
+layout: post
+title: LiveView Design Patterns: LiveComponent and the Single Responsibility Principle
+excerpt: >
+  It's easy to end up with an overly complex LiveView that houses lots of business rules and responsibilities. We can use `Phoenix.LiveComponent` to build a LiveView feature that is clean, maintainable and adherent to the Single Responsibility Principle.
+---
+
+## LiveView Can Get Messy  
+
+As LiveView becomes a more established technology, we naturally find ourselves leveraging to back more and more complex features. If we're not careful, this can lead to "fat controller syndrome"––live views that are jam packed with complex business logic and disparate responsibilities, just like the classic "fat Rails controller".
```
Comment (gemantzu) [inline _posts/2019-12-12-live-view-live-component.md]: leveraging to back?

File: _posts/2019-12-12-live-view-live-component.md [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+How can we write live views that are easy to reason about and maintain and adhere to common design principles like SRP?
```
Comment (gemantzu) [inline _posts/2019-12-12-live-view-live-component.md]: rephrase to remove one and :)

File: _posts/2019-12-12-live-view-live-component.md [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+
+One way to achieve this goal is to leverage the `Phoenix.LiveComponent` behaviour.
+
+## Introducing `Phoenix.LiveComponent`
+
+Components are modules that use the `Phoenix.LiveComponent` behaviour. This behaviour provides
+
+> ...a mechanism to compartmentalize state, markup, and events in LiveView. [*](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveComponent.html)
```
Comment (gemantzu) [inline _posts/2019-12-12-live-view-live-component.md]: An asterisk will probably difficult to click on, more so on a cell phone. You should change it to something with more text :)

File: _posts/2019-12-12-live-view-live-component.md [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+      ) do
+    {:noreply, assign(socket, :search, %Search{query: query, value: search.value})}
+  end
+end
+```
+
+These callbacks ensure two things for us:
+
+* The correct radio button is marked as "selected" when a user chooses a new search query type option.
+* The search form input's `placeholder` attribute is correctly updated to reflected the selected query type:
```
Comment (gemantzu) [inline _posts/2019-12-12-live-view-live-component.md]: to reflect

File: _posts/2019-12-12-live-view-live-component.md [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+The feature we're designing requires us to populate the URI's query params in the browser's URL bar when the user submits the search form. This allows users to share a link with the results of a particular search.
+
+In order to achieve this, we can reach for the `live_redirect/2` function. This will take advantage of the browser's `pushState` API to change the page navigation without actually sending a web request. Instead, our live view's `handle_params/3` callback function will invoked, allowing us to respond by searching for the appropriate messages and updating the live view socket's state with those messages.
+
+But wait! The `live_redirect/2` function is sadly not available from within component since the `Phoenix.LiveComponent` behaivour does not implement a `handle_params/3` function. Luckily for us, however, the parent live view and the component share a process. That means that calling `self()` from within the component returns a PID that *is the same PID as that parent live view process*. So, from within our component we can `send` a message to `self()` and handle that message in the parent live view.
```
Comment (gemantzu) [inline _posts/2019-12-12-live-view-live-component.md]: behaviour

**Comment by gemantzu** [comment]
A couple of typos I found.
Again, I feel we are missing an opportunity here. @SophieDeBenedetto does a great job with these blog posts, but I think they would have a better place as an actual lesson. Even if we do not have yet a Phoenix Section, we could make a LiveView section where these blog posts could migrate to.
Also, I think having a repo to work on for someone that reads this post would be awesome.
All that aside, great work again @SophieDeBenedetto , I don't know how you do it each time but thank you :)

**Comment by SophieDeBenedetto** [comment]
> A couple of typos I found.
> Again, I feel we are missing an opportunity here. @SophieDeBenedetto does a great job with these blog posts, but I think they would have a better place as an actual lesson. Even if we do not have yet a Phoenix Section, we could make a LiveView section where these blog posts could migrate to.
> Also, I think having a repo to work on for someone that reads this post would be awesome.
> All that aside, great work again @SophieDeBenedetto , I don't know how you do it each time but thank you :)

Thanks for the feedback @gemantzu! I'll take a pass at these edits later today. I'm also totally open to doing a LiveView section of lessons, but I know we're gone back and forth in the past on whether we would include Phoenix content at all. Maybe I'll move this convo to the Slack channel and see what people think. Thanks again!

**Comment by SophieDeBenedetto** [comment]
@gemantzu addressed your comments!

**Comment by SophieDeBenedetto** [comment]
Thanks @gemantzu! I just updated the date to tomorrow so can publish then provided you can give a quick re-approval!

**Pull request merged**